### PR TITLE
Add ColorMatrix.scala, ImageFilter.scala #13

### PR DESCRIPTION
**Pull request by theolundqvist**
ColorMatrix:
- Wrap BufferedImage 
- simplifies editing and getting pixels
- method to return matrix of colors

PixelWindow:
- display ColorMatrix, Matrix of Colors, BufferedImage.
- take screenshot of whole window and section, return as ColorMatrix

ImageFilter translated from java to scala

Added a fun test.

#13

### CODE REVIEW

**Comment by theolundqvist** [comment]
image

**Comment by bjornregnell** [comment]
Nice! According to our discussion: rename `ColorMatrix` to `Image` and simplify the API to not directly use java...BufferedImage as discussed in https://github.com/lunduniversity/introprog-scalalib/issues/13#issuecomment-889797910

**Comment by theolundqvist** [comment]
@bjornregnell Im satisfied with how it turned out. Fast and powerful methods and ease of use

File: src/main/scala/introprog/IO.scala [hunk 1]
```diff
@@ -139,23 +139,76 @@ object IO {
...
-    * Loads an image from file.
+    * Load image from file.
     *
     * @param fileName the path the image that will be loaded.
-    * @return BufferedImage.
     * */
-  def loadImage(fileName: String): java.awt.image.BufferedImage = 
-    javax.imageio.ImageIO.read(new java.io.File(fileName))
-  
+  def loadImage(fileName: String): Option[Image] =
```
Comment (bjornregnell) [inline src/main/scala/introprog/IO.scala:146]: The problem with returning Option is that the user never gets to see what went wrong. We could use Either[Throwable, Image] but ]I'd rather return just an Image and let it just fail with exception for simplicitly. Then the user does not have to know about neither Either nor Option. And the user can always wrap it in a Try with the same effect as here.

File: src/main/scala/introprog/IO.scala [hunk 2]
```diff
@@ -139,23 +139,76 @@ object IO {
...
+  /**
+  * Save `img` to file as `JPEG` with a compression ratio of 0.75.
+  * Restore color of transparent pixels.
+  * @param img the image to save.
+  * @param fileName the path to save the image to, `path/file.jpg` or just `path/file`.
+  * */
+  def saveJPEG(img: Image, fileName: String) : Unit = 
+    import javax.imageio.ImageIO
+    import java.io.File
+    if !ImageIO.write(img.underlying, "jpg", File(if fileName.endsWith(".jpg") then fileName else s"$fileName.jpg")) then
```
Comment (bjornregnell) [inline src/main/scala/introprog/IO.scala:196]: This is difficult to read. Better to first capture the appropriate file name in a val e.g. `val f = ....`, thus avoiding nested ifs.

File: src/main/scala/introprog/IO.scala [hunk 3]
```diff
@@ -139,23 +139,76 @@ object IO {
...
+  /**
+  * Save `img` to file as `PNG`.
+  *
+  * @param img the image to save.
+  * @param fileName the path to save the image to, `path/file.png` or just `path/file`.
+  * */
+  def savePNG(img: Image, fileName: String) : Unit = 
+    import javax.imageio.ImageIO
+    import java.io.File
+    if !ImageIO.write(img.underlying, "png", File(if fileName.endsWith(".png") then fileName else s"$fileName.png")) then
```
Comment (bjornregnell) [inline src/main/scala/introprog/IO.scala:208]: Same comment as previous method.

File: src/main/scala/introprog/Image.scala [hunk 4]
```diff
@@ -0,0 +1,41 @@
+package introprog
+
+class Image (val underlying: java.awt.image.BufferedImage):
+  import java.awt.Color
+  import java.awt.image.BufferedImage
+
+  def apply(x: Int, y: Int): Color = new Color(underlying.getRGB(x, y)) //getRGB ger en Int
```
Comment (bjornregnell) [inline src/main/scala/introprog/Image.scala:7]: Skip new here as possible in Scala 3.
Comments in english.

File: src/main/scala/introprog/Image.scala [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  def toMatrix: Array[Array[Color]] = 
+    val xs: Array[Array[Color]] = Array.ofDim(width, height)
+    for x <- 0 until width; y <- 0 until height do
+      xs(x)(y) = apply(x, y)
+    xs
+
+  /** Copy subsection of image defined by top left corner `(x, y)` and `(width, height)`.*/
+  def subsection(x: Int, y: Int, width: Int, height: Int): Image =   
+    val bi = BufferedImage(width, height, underlying.getType)
+    bi.createGraphics().drawImage(underlying, 0, 0, width, height, x, y, x+width, y+height, null)
```
Comment (bjornregnell) [inline src/main/scala/introprog/Image.scala:24]: spaces around +

File: src/main/scala/introprog/Image.scala [hunk 6]
```diff
@@ -0,0 +1,41 @@
+package introprog
+
+class Image (val underlying: java.awt.image.BufferedImage):
+  import java.awt.Color
+  import java.awt.image.BufferedImage
+
+  def apply(x: Int, y: Int): Color = new Color(underlying.getRGB(x, y)) //getRGB ger en Int
+  def update(x: Int, y: Int, c: Color): Unit = underlying.setRGB(x, y, c.getRGB)
```
Comment (bjornregnell) [inline src/main/scala/introprog/Image.scala:8]: Missing doc comment.

File: src/main/scala/introprog/Image.scala [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+  def update(f: (Int, Int) => Color): Unit =
```
Comment (bjornregnell) [inline src/main/scala/introprog/Image.scala:10]: Missing doc comment.

File: src/main/scala/introprog/ImageFilter.scala [hunk 8]
```diff
@@ -0,0 +1,82 @@
+
+
+package introprog
+
+
+/**
+ * Superklassen till alla filterklasser.
+ * 
+ * @version 1.3 (2021-07-25) översättning från java
```
Comment (bjornregnell) [inline src/main/scala/introprog/ImageFilter.scala:9]: Unsure if this should be a version number here as it creates confusion with the introporg version number. 
So remove version numbers but keep dates and authors.

**Comment by bjornregnell** [comment]
> @bjornregnell Im satisfied with how it turned out. Fast and powerful methods and ease of use

Yeah this will be great. I have some change proposals in the review comments attached to the code. Let me know if you don't agree and we can discuss uptions.

Also,  the ImageFilter class is perhaps not exactly Scala-idomatic; it feels more like Java in Scala :) but let's leave like that for now, and we can discuss options when after the lab solution as been re-implemented with the new api.

File: src/main/scala/introprog/Image.scala [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    Image(bi)
+
+  /** Copy image and scale to `(width, height)`.*/
+  def scaled(width: Int, height: Int): Image = 
+    val bi = BufferedImage(width, height, underlying.getType)
+    bi.createGraphics().drawImage(underlying, 0, 0, width, height, null)
+    Image(bi)
+  
+  /** Copy image and change image type ex. BufferedImage.TYPE_INT_RGB*/
+  def toImageType(imageType: Int): Image =
```
[the remark this answers is not in the document]
Comment (theolundqvist) [reply src/main/scala/introprog/Image.scala:34]: Could do either of these alternatives
1. seperate into `toRGB` and `toARGB` or `add/removeAlphaLayer`.
2. remove method and use underlying bufferedimage in saveJPEG to remove alphalayer. If the user want to change imagetype they would have to use bufferedImage.
3. add `@param imageType` with all possible imagetypes for easy reading in editor

**Comment by theolundqvist** [comment]
> > @bjornregnell Im satisfied with how it turned out. Fast and powerful methods and ease of use
> 
> Yeah this will be great! I have some change proposals in the review comments attached to the code. Let me know if you don't agree and we can discuss options.
> 
> Also, the ImageFilter class is perhaps not exactly Scala-idomatic; it feels more like Java in Scala :) but let's leave like that for now, and we can discuss options when after the lab solution as been re-implemented with the new api.

I'm not sure how the methods in ImageFilter works so it was hard to do anything but straight translation of the code.

File: src/main/scala/introprog/Image.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (theolundqvist) [reply src/main/scala/introprog/Image.scala:34]: `setTransparency`

[the remark this answers is not in the document]
Comment (theolundqvist) [reply src/main/scala/introprog/Image.scala:34]: withAlpha
withoutAlpha
hasAlpha

decided

[the remark this answers is not in the document]
Comment (bjornregnell) [reply src/main/scala/introprog/Image.scala:34]: :+1:

**Comment by theolundqvist** [comment]
Requested changes should now be done. @bjornregnell

**Review by bjornregnell — APPROVED** [review]

**Comment by bjornregnell** [comment]
I have just 
* made a minor fix https://github.com/lunduniversity/introprog-scalalib/commit/90ebfb8c838d591f0c8d1b5af1d8596a639d57b2 
* and removed ImageFilter from here as this should be in the photo project assignment instead.
https://github.com/lunduniversity/introprog-scalalib/commit/18532f7a7237fa3e35b32ca2141727776291d0f3

**Pull request merged**